shared/src/main/scala/mlscript/ConstraintSolver.scala

```diff
@@ -325,139 +340,106 @@ class ConstraintSolver extends NormalForms { self: Typer =>
-          first = false
-          msg"${msgHead}from ${p.desc}:" -> p.loco
+      val originProvHints = originProvList.map { case (nme, l) => 
+        val msgHead =
+          if (first)
+                msg"Note: ${l.desc} $nme"
+          else  msg"      ${l.desc} $nme"
+        first = false
+        msg"${msgHead} is defined at: " -> l.loco 
       }
       
-      val originProvList = cctx.flatMap{ subCtx =>
-        subCtx.iterator.flatMap { case (l,r) => 
-          // We only keep one origin prov for each subCtx.
-          val considered = List(l optionIf (_.prov.isOrigin), r optionIf (_.prov.isOrigin)).flatten
-          considered
-        }.nextOption()
-      }
-
-      first = true
-      val originProvHints = originProvList.map { l => 
-        val msgHead = if (first) msg"Note: " else msg""
-          first = false
-          msg"${msgHead}${l.prov.desc.capitalize} is defined at: " -> l.prov.loco 
-      }
-
-      val msgs: Ls[Message -> Opt[Loc]] = List(
+      val msgs: Ls[Message -> Opt[Loc]] = Ls[Ls[Message -> Opt[Loc]]](
         msg"Type mismatch in ${prov.desc}:" -> prov.loco :: Nil,
         msg"expression of type `${lhs.expPos}` $error" ->
           (if (lhsProv.loco === prov.loco) N else lhsProv.loco) :: Nil,
-        tighestRelevantFailure.map { case (l, r, isSameType) =>
-          // Note: used to have `val isSameType = l.unwrapProxies === lhs.unwrapProxies`
-          //  which was only an approximation, and considered things like `?a | int` not the same as `int`.
-          val expTyMsg =
-            if (isSameType) msg" with expected type `${r.expNeg}`"
-            // ^ This used to say "of expected type", but in fact we're describing a term with the wrong type;
-            //   the expected type is not its type so that was not a good formulation.
-            else msg" of type `${l.expPos}`"
-          val lunw = l.unwrapProxies
-          lazy val fail = (l, r) match {
-            case (RecordType(fs0), RecordType(fs1)) =>
-              (fs0.map(_._1).toSet -- fs1.map(_._1).toSet).headOption.fold(doesntMatch(r)) { n1 =>
-                doesntHaveField(n1.name)
-              }
-            case (_, FunctionType(_, _))
-              if !lunw.isInstanceOf[FunctionType]
-              && !lunw.isInstanceOf[TypeVariable]
-              => msg"is not a function"
-            case (_, RecordType((n, _) :: Nil))
-              if !lunw.isInstanceOf[RecordType]
-              && !lunw.isInstanceOf[TypeVariable]
-              => doesntHaveField(n.name)
-            case _ => doesntMatch(r)
-          }
-          msg"but it flows into ${l.prov.desc}$expTyMsg" -> l.prov.loco ::
-          (if (isSameType) Nil else msg"which $fail" -> N :: Nil)
+        tighestRelevantFailure.map { l =>
+          val expTyMsg = msg" with expected type `${rhs.expNeg}`"
+          msg"but it flows into ${l.prov.desc}$expTyMsg" -> l.prov.loco :: Nil
         }.toList.flatten,
-        constraintProvenanceHints,
```

Comment:
The variable constraintProvenanceHints containing the updated content can probably be kept to keep the logic used with originProvHints and detailedContext

Reply:
Sure, good idea. I've made the change in the follow-up PR though, to avoid conflicts: https://github.com/hkust-taco/mlscript/pull/74/commits/ea98eb79ea388996d4e77c9418091826fe05d7eb.